src/gsAssembler/gsAssembler.h

```diff
@@ -34,6 +34,236 @@
+        case 1:
+            secDers(0, 0) = hessian(0, 0);
+            break;
+        case 2:
+            secDers(0, 0) = hessian(0, 0);
+            secDers(0, 1) = hessian(1, 1);
+            secDers(0, 2) = (hessian(1, 0) + hessian(0, 1)) / 2.0;
+            break;
+        case 3:
+            secDers(0, 0) = hessian(0, 0);
+            secDers(0, 1) = hessian(1, 1);
+            secDers(0, 2) = hessian(2, 2);
+            secDers(0, 3) = (hessian(0, 1) + hessian(1, 0)) / 2.0;
+            secDers(0, 4) = (hessian(0, 2) + hessian(2, 0)) / 2.0;
+            secDers(0, 5) = (hessian(1, 2) + hessian(2, 1)) / 2.0;
+            break;
+        default:
+            break;
+    }
+}
+
+template<typename T>
+void secDerToTensor(Eigen::DenseBase<Eigen::Map<const Eigen::Matrix<double, -1, -1>, 0, Eigen::Stride<0, 0> > >::ConstColXpr & secDers,
+                    gsMatrix<T> * a,
+                    int parDim, int geoDim)
+{
+    static const int dim = parDim * (parDim + 1) / 2;
+    for (int i = 0; i < geoDim; ++i)
+        secDerToHessian<T>(secDers.segment(i * dim, dim), a[i], parDim);
+}
+
+template <class T>
+void transformDeriv2Hgrad(const gsMapData<T> & md, 
+                          index_t              k,
+                          const gsMatrix<T> &  funcGrad,
+                          const gsMatrix<T> &  funcSecDir,
+                          gsMatrix<T> &        result)
+{
+    //todo: check me
+    static const int ParDim = md.dim.first;
+    static const int GeoDim = md.dim.second;
+    GISMO_ASSERT(
+        (ParDim == 1 && (GeoDim == 1 || GeoDim == 2 || GeoDim == 3))
+        || (ParDim == 2 && (GeoDim == 2 || GeoDim == 3))
+        || (ParDim == 3 && GeoDim == 3), "No implementation for this case");
+
+    // important sizes
+    static const int parSecDirSize = ParDim * (ParDim + 1) / 2;
+    static const int fisSecDirSize = GeoDim * (GeoDim + 1) / 2;
+
+    // allgrads
+    const index_t numGrads = funcGrad.rows() / ParDim;
+
+    result.resize(numGrads, fisSecDirSize);
+
+    typename gsMatrix<T>::constRef JM1 = md.jacobian(k).inverse().transpose();  // m_jacInvs.template block<GeoDim,ParDim>(0, k*ParDim);
+    typename gsMatrix<T>::constRef JMT = md.jacobian(k).inverse();              // m_jacInvs.template block<GeoDim,ParDim>(0, k*ParDim).transpose();
+
+    gsDebugIf(ParDim != 2, ParDim);
+    gsDebugIf(ParDim != 2, JM1);
+    gsDebugIf(ParDim != 2, JMT);
+
+    // First part: J^-T H(u) J^-1
+    gsMatrix<T> parFuncHessian;
+    for (index_t i = 0; i < numGrads; ++i)
+    {
+        secDerToHessian<T>(funcSecDir.block(i * parSecDirSize, k, parSecDirSize, 1), parFuncHessian, ParDim);
+        hessianToSecDer<T>(JM1 * parFuncHessian * JMT, result.row(i), GeoDim);
+    }
+
+    // Second part: sum_i[  J^-T H(G_i) J^-1 ( e_i^T J^-T grad(u) ) ]
+    const gsAsConstMatrix<T, -1, -1>  & secDer = md.deriv2(k);
+#ifndef _MSC_VER
+    gsMatrix<T> DDG[GeoDim]; // Each matrix is the Hessian of a component of the Geometry
+#else
```

Comment:
can you use the STACK_ARRAY macro here?

certainly, this is not allowed in standard c++ because GeoDim cannot be known at compile time.

in every case, it would be good to encapsulate such constructs and give the implementation in gsCore/gsMemory.h

If this is done on the heap, consider to use memory::unique_ptr

Reply:
nope - doesn't work.

./biharmonic_example 
Assembling...
double free or corruption (out)
Aborted (core dumped)

For the moment we should keep it as it is..

Reply:
maybe we should move all of this functions to gsFuncData or gsMapData.. than we could use template parameters again - inverse() is way faster in such a case (knows ParDim at compile time), or?